File: ui/src/stores/project.js
import { writable, derived } from "svelte/store";
import { invoke } from "../lib/tauri.js";

export const project = writable(null);
export const projectPath = writable(null);

export const kozijnen = derived(project, ($project) =>
  $project ? $project.kozijnen : []
);

export async function loadProject() {
  try {
    const p = await invoke("get_project");
    project.set(p);
    return p;
  } catch (e) {
    console.error("Laden project mislukt:", e);
    return null;
  }
}

export async function newProject(name, number) {
  const p = await invoke("new_project", { name, number });
  project.set(p);
  projectPath.set(null);
  return p;
}

export async function openProject(filePath) {
  const p = await invoke("open_project", { filePath });
  project.set(p);
  projectPath.set(filePath);
  return p;
}

export async function saveProject(filePath) {
  await invoke("save_project", { filePath });
  projectPath.set(filePath);
}

export async function refreshProject() {
  return loadProject();
}

File: src-tauri/src/commands/project.rs
use crate::state::AppState;
use ofs_core::kozijn::Project;
use tauri::State;

#[tauri::command]
pub fn new_project(
    state: State<'_, AppState>,
    name: String,
    number: String,
) -> Result<Project, String> {
    let project = Project::new(&name, &number);
    let mut current = state.project.lock().map_err(|e| e.to_string())?;
    *current = project.clone();
    let mut path = state.project_path.lock().map_err(|e| e.to_string())?;
    *path = None;
    Ok(project)
}

#[tauri::command]
pub fn get_project(state: State<'_, AppState>) -> Result<Project, String> {
    let project = state.project.lock().map_err(|e| e.to_string())?;
    Ok(project.clone())
}

#[tauri::command]
pub fn open_project(state: State<'_, AppState>, file_path: String) -> Result<Project, String> {
    let contents = std::fs::read_to_string(&file_path).map_err(|e| e.to_string())?;
    let project: Project = serde_json::from_str(&contents).map_err(|e| e.to_string())?;
    let mut current = state.project.lock().map_err(|e| e.to_string())?;
    *current = project.clone();
    let mut path = state.project_path.lock().map_err(|e| e.to_string())?;
    *path = Some(file_path);
    Ok(project)
}

#[tauri::command]
pub fn save_project(state: State<'_, AppState>, file_path: String) -> Result<(), String> {
    let project = state.project.lock().map_err(|e| e.to_string())?;
    let json = serde_json::to_string_pretty(&*project).map_err(|e| e.to_string())?;
    std::fs::write(&file_path, json).map_err(|e| e.to_string())?;
    drop(project);
    let mut path = state.project_path.lock().map_err(|e| e.to_string())?;
    *path = Some(file_path);
    Ok(())
}
